File: packages/js/datagen.js
// Reference implementation of the vector generated-data algorithm and its
// derived digest values. Normative prose: the repository README, "Generated data".
import { createHash } from 'node:crypto'

// block(i) = SHA256(UTF8(seed) || BE64(i)); stream = block(0) || block(1) || ...
function prng (seed, size) {
  const out = Buffer.alloc(size)
  const seeded = createHash('sha256').update(Buffer.from(seed, 'utf8'))
  const counter = Buffer.alloc(8)
  for (let off = 0, i = 0n; off < size; off += 32, i++) {
    counter.writeBigUInt64BE(i)
    const block = seeded.copy().update(counter).digest()
    block.copy(out, off, 0, Math.min(32, size - off))
  }
  return out
}

function pattern (patternBytes, size) {
  if (patternBytes.length === 0) throw new Error('empty pattern')
  const out = Buffer.alloc(size)
  for (let off = 0; off < size; off += patternBytes.length) {
    patternBytes.copy(out, off, 0, Math.min(patternBytes.length, size - off))
  }
  return out
}

/**
 * Materialize one named dataset from a vector's `data` map.
 * @param {Record<string, object>} specs the vector's `data` map
 * @param {string} name
 * @returns {Buffer}
 */
export function generate (specs, name) {
  const spec = specs[name]
  if (!spec) throw new Error(`unknown dataset: ${name}`)
  switch (spec.kind) {
    case 'prng':
      return prng(spec.seed, spec.size)
    case 'pattern':
      return pattern(
        spec.pattern !== undefined ? Buffer.from(spec.pattern, 'utf8') : Buffer.from(spec.patternBase64, 'base64'),
        spec.size
      )
    case 'slice': {
      const parent = specs[spec.of]
      if (!parent) throw new Error(`slice '${name}' references unknown dataset '${spec.of}'`)
      if (parent.kind === 'slice') throw new Error(`slice '${name}' references slice '${spec.of}' (chained slices are not allowed)`)
      const base = generate(specs, spec.of)
      if (spec.offset + spec.length > base.length) {
        throw new Error(`slice '${name}' [${spec.offset}, ${spec.offset + spec.length}) exceeds '${spec.of}' size ${base.length}`)
      }
      return base.subarray(spec.offset, spec.offset + spec.length)
    }
    default:
      throw new Error(`unknown data kind: ${spec.kind}`)
  }
}

function makeCrc32Table (poly) {
  const table = new Uint32Array(256)
  for (let n = 0; n < 256; n++) {
    let c = n
    for (let k = 0; k < 8; k++) c = c & 1 ? (c >>> 1) ^ poly : c >>> 1
    table[n] = c >>> 0
  }
  return table
}

const CRC32_TABLE = makeCrc32Table(0xEDB88320)
const CRC32C_TABLE = makeCrc32Table(0x82F63B78)

function crc32 (buf, table) {
  let c = 0xFFFFFFFF
  for (let i = 0; i < buf.length; i++) c = table[(c ^ buf[i]) & 0xFF] ^ (c >>> 8)
  return (c ^ 0xFFFFFFFF) >>> 0
}

// CRC-64/NVME: reflected poly 0x9A6C9329AC4BC9B5, init/xorout all-ones.
const CRC64_TABLE = (() => {
  const poly = 0x9A6C9329AC4BC9B5n
  const table = new BigUint64Array(256)
  for (let n = 0; n < 256; n++) {
    let c = BigInt(n)
    for (let k = 0; k < 8; k++) c = c & 1n ? (c >> 1n) ^ poly : c >> 1n
    table[n] = c
  }
  return table
})()

function crc64nvme (buf) {
  let c = 0xFFFFFFFFFFFFFFFFn
  for (let i = 0; i < buf.length; i++) {
    c = CRC64_TABLE[Number((c ^ BigInt(buf[i])) & 0xFFn)] ^ (c >> 8n)
  }
  return c ^ 0xFFFFFFFFFFFFFFFFn
}

function u32ToBase64 (v) {
  const b = Buffer.alloc(4)
  b.writeUInt32BE(v)
  return b.toString('base64')
}

function u64ToBase64 (v) {
  const b = Buffer.alloc(8)
  b.writeBigUInt64BE(v)
  return b.toString('base64')
}

/** The fields available as `${data.<name>.<field>}` placeholders. */
export const DERIVED_FIELDS = Object.freeze([
  'size', 'md5', 'etag', 'sha256', 'sha256B64', 'sha1B64', 'crc32B64', 'crc32cB64', 'crc64nvmeB64'
])

/**
 * Compute a derived string value of a dataset (what a `${data.<name>.<field>}`
 * placeholder resolves to).
 * @param {Record<string, object>} specs the vector's `data` map
 * @param {string} name
 * @param {string} field one of DERIVED_FIELDS
 * @returns {string}
 */
export function derived (specs, name, field) {
  const bytes = generate(specs, name)
  switch (field) {
    case 'size': return String(bytes.length)
    case 'md5': return createHash('md5').update(bytes).digest('hex')
    case 'etag': return `"${createHash('md5').update(bytes).digest('hex')}"`
    case 'sha256': return createHash('sha256').update(bytes).digest('hex')
    case 'sha256B64': return createHash('sha256').update(bytes).digest('base64')
    case 'sha1B64': return createHash('sha1').update(bytes).digest('base64')
    case 'crc32B64': return u32ToBase64(crc32(bytes, CRC32_TABLE))
    case 'crc32cB64': return u32ToBase64(crc32(bytes, CRC32C_TABLE))
    case 'crc64nvmeB64': return u64ToBase64(crc64nvme(bytes))
    default: throw new Error(`unknown derived data field: ${field}`)
  }
}

File: packages/rust/src/datagen.rs
//! Rust port of the vector generated-data reference: materialize a vector's
//! named datasets and compute the derived digest strings that
//! `${data.<name>.<field>}` placeholders resolve to.

use std::collections::BTreeMap;
use std::fmt;

use base64::engine::general_purpose::STANDARD as BASE64;
use base64::Engine as _;
use crc::Crc;
use md5::Md5;
use sha1::Sha1;
use sha2::{Digest as _, Sha256};

use crate::model::DataSpec;

/// The fields available as `${data.<name>.<field>}` placeholders.
pub const DERIVED_FIELDS: [DerivedField; 9] = [
    DerivedField::Size,
    DerivedField::Md5,
    DerivedField::Etag,
    DerivedField::Sha256,
    DerivedField::Sha256B64,
    DerivedField::Sha1B64,
    DerivedField::Crc32B64,
    DerivedField::Crc32cB64,
    DerivedField::Crc64NvmeB64,
];

/// A derived digest field.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DerivedField {
    Size,
    Md5,
    Etag,
    Sha256,
    Sha256B64,
    Sha1B64,
    Crc32B64,
    Crc32cB64,
    Crc64NvmeB64,
}

impl DerivedField {
    /// The field name as written in placeholders (`crc32cB64`, ...).
    pub fn as_str(&self) -> &'static str {
        match self {
            DerivedField::Size => "size",
            DerivedField::Md5 => "md5",
            DerivedField::Etag => "etag",
            DerivedField::Sha256 => "sha256",
            DerivedField::Sha256B64 => "sha256B64",
            DerivedField::Sha1B64 => "sha1B64",
            DerivedField::Crc32B64 => "crc32B64",
            DerivedField::Crc32cB64 => "crc32cB64",
            DerivedField::Crc64NvmeB64 => "crc64nvmeB64",
        }
    }
}

impl std::str::FromStr for DerivedField {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Error> {
        DERIVED_FIELDS
            .into_iter()
            .find(|f| f.as_str() == s)
            .ok_or_else(|| Error::UnknownField(s.to_string()))
    }
}

/// Datagen error.
#[derive(Debug)]
pub enum Error {
    UnknownDataset(String),
    UnknownField(String),
    ChainedSlice { name: String, of: String },
    SliceOutOfRange { name: String, of: String },
    BadPattern(String),
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Error::UnknownDataset(name) => write!(f, "unknown dataset: {name}"),
            Error::UnknownField(field) => write!(f, "unknown derived data field: {field}"),
            Error::ChainedSlice { name, of } => {
                write!(f, "slice {name:?} references slice {of:?} (chained slices are not allowed)")
            }
            Error::SliceOutOfRange { name, of } => {
                write!(f, "slice {name:?} exceeds bounds of {of:?}")
            }
            Error::BadPattern(name) => write!(f, "dataset {name:?} has an invalid pattern"),
        }
    }
}

impl std::error::Error for Error {}

// block(i) = SHA256(UTF8(seed) || BE64(i)); stream = block(0) || block(1) || ...
fn prng(seed: &str, size: u64) -> Vec<u8> {
    let size = size as usize;
    let mut out = vec![0u8; size];
    let mut i: u64 = 0;
    let mut off = 0usize;
    while off < size {
        let mut h = Sha256::new();
        h.update(seed.as_bytes());
        h.update(i.to_be_bytes());
        let block = h.finalize();
        let n = usize::min(32, size - off);
        out[off..off + n].copy_from_slice(&block[..n]);
        off += 32;
        i += 1;
    }
    out
}

fn pattern(name: &str, pattern_bytes: &[u8], size: u64) -> Result<Vec<u8>, Error> {
    if pattern_bytes.is_empty() {
        return Err(Error::BadPattern(name.to_string()));
    }
    let size = size as usize;
    let mut out = vec![0u8; size];
    let mut off = 0usize;
    while off < size {
        let n = usize::min(pattern_bytes.len(), size - off);
        out[off..off + n].copy_from_slice(&pattern_bytes[..n]);
        off += pattern_bytes.len();
    }
    Ok(out)
}

/// Materialize one named dataset from a vector's `data` map.
pub fn generate(specs: &BTreeMap<String, DataSpec>, name: &str) -> Result<Vec<u8>, Error> {
    let spec = specs
        .get(name)
        .ok_or_else(|| Error::UnknownDataset(name.to_string()))?;
    match spec {
        DataSpec::Prng(d) => Ok(prng(&d.seed, d.size)),
        DataSpec::Pattern(d) => match (&d.pattern, &d.pattern_base64) {
            (Some(p), _) => pattern(name, p.as_bytes(), d.size),
            (None, Some(b64)) => {
                let raw = BASE64
                    .decode(b64)
                    .map_err(|_| Error::BadPattern(name.to_string()))?;
                pattern(name, &raw, d.size)
            }
            (None, None) => Err(Error::BadPattern(name.to_string())),
        },
        DataSpec::Slice(d) => {
            let parent = specs
                .get(&d.of)
                .ok_or_else(|| Error::UnknownDataset(d.of.clone()))?;
            if matches!(parent, DataSpec::Slice(_)) {
                return Err(Error::ChainedSlice {
                    name: name.to_string(),
                    of: d.of.clone(),
                });
            }
            let base = generate(specs, &d.of)?;
            let end = d
                .offset
                .checked_add(d.length)
                .ok_or_else(|| Error::SliceOutOfRange {
                    name: name.to_string(),
                    of: d.of.clone(),
                })?;
            if end > base.len() as u64 {
                return Err(Error::SliceOutOfRange {
                    name: name.to_string(),
                    of: d.of.clone(),
                });
            }
            Ok(base[d.offset as usize..end as usize].to_vec())
        }
    }
}

const CRC32: Crc<u32> = Crc::<u32>::new(&crc::CRC_32_ISO_HDLC);
const CRC32C: Crc<u32> = Crc::<u32>::new(&crc::CRC_32_ISCSI);
const CRC64NVME: Crc<u64> = Crc::<u64>::new(&crc::CRC_64_NVME);

/// Compute the string a `${data.<name>.<field>}` placeholder resolves to.
pub fn derived(
    specs: &BTreeMap<String, DataSpec>,
    name: &str,
    field: DerivedField,
) -> Result<String, Error> {
    let bytes = generate(specs, name)?;
    Ok(match field {
        DerivedField::Size => bytes.len().to_string(),
        DerivedField::Md5 => hex(&Md5::digest(&bytes)),
        DerivedField::Etag => format!("\"{}\"", hex(&Md5::digest(&bytes))),
        DerivedField::Sha256 => hex(&Sha256::digest(&bytes)),
        DerivedField::Sha256B64 => BASE64.encode(Sha256::digest(&bytes)),
        DerivedField::Sha1B64 => BASE64.encode(Sha1::digest(&bytes)),
        DerivedField::Crc32B64 => BASE64.encode(CRC32.checksum(&bytes).to_be_bytes()),
        DerivedField::Crc32cB64 => BASE64.encode(CRC32C.checksum(&bytes).to_be_bytes()),
        DerivedField::Crc64NvmeB64 => BASE64.encode(CRC64NVME.checksum(&bytes).to_be_bytes()),
    })
}

fn hex(bytes: &[u8]) -> String {
    let mut s = String::with_capacity(bytes.len() * 2);
    for b in bytes {
        s.push_str(&format!("{b:02x}"));
    }
    s
}
